src/BiDirectionalLinkedListImp.h

```diff
@@ -0,0 +1,152 @@
+//
+// Created by shvmo on 02/05/2020.
+//
+
+#ifndef DATASTRUCTURESWET1_BIDIRECTIONALLINKEDLISTIMP_H
+#define DATASTRUCTURESWET1_BIDIRECTIONALLINKEDLISTIMP_H
+
+#include <cassert>
+#include "BiDirectionalLinkedList.h"
+
+using std::make_shared;
+
+template<class Type>
+BiDirectionalLinkedList<Type>::BiDirectionalLinkedList() = default;
+
+template<class Type>
+BiDirectionalLinkedList<Type>::~BiDirectionalLinkedList() = default;
+
+template<class Type>
+shared_ptr<BiDirectionalNode<Type>> BiDirectionalLinkedList<Type>::getFirst() const {
+    return this->first;
+}
+
+template<class Type>
+BiDirectionalNode<Type> *BiDirectionalLinkedList<Type>::getFirstPointer() const {
+    return this->first.get();
+}
+
+template<class Type>
+shared_ptr<BiDirectionalNode<Type>> BiDirectionalLinkedList<Type>::getLast() const {
+    return this->last;
+}
+
+template<class Type>
+void BiDirectionalLinkedList<Type>::insert(const Type &newValue) {
+    if (this->first == nullptr) {
+        this->first = make_shared<BiDirectionalNode<Type>>();
+
+        this->first->setNext(nullptr);
+        this->first->setPrevious(nullptr);
+        this->first->setValue(newValue);
+
+        this->last = this->first;
+    } else {
+        shared_ptr<BiDirectionalNode<Type>> newLast = make_shared<BiDirectionalNode<Type>>();
+
+        newLast->setNext(nullptr);
+        newLast->setPrevious(last);
+        newLast->setValue(newValue);
+
+        last->setNext(newLast);
+
+        this->last = newLast;
+    }
+}
+
+template<class Type>
+shared_ptr<BiDirectionalNode<Type>> BiDirectionalLinkedList<Type>::findNodeByValue(const Type &valueToFind) const {
+    for (shared_ptr<BiDirectionalNode<Type>> current = this->first; current != nullptr; current = current->getNext()) {
+        if (valueToFind == current->getValue()) {
+            return current;
+        }
+    }
+}
+
+template<class Type>
+void BiDirectionalLinkedList<Type>::deleteNodeByValue(const Type &valueToFind) {
+    for (shared_ptr<BiDirectionalNode<Type>> current = this->first; current != nullptr; current = current->getNext()) {
+        if (valueToFind == current->getValue()) {
+            remove(current);
+
+            break;
+        }
+    }
+}
+
+template<class Type>
+void BiDirectionalLinkedList<Type>::printList(ostream &os) const {
+    for (shared_ptr<BiDirectionalNode<Type>> current = this->first; current != nullptr; current = current->getNext()) {
```

Comment:
why not run from `this->first` to `this->last` and then change
`os << std::endl;`
to
`os << this->last.get() << std::endl;`

Reply:
same for `printListReverse()`

Reply:
ok now isnstead of "dangerous" the only reason for this comment is Readability.
so change it only if you want to. otherwise resolve.